### PR TITLE
feature(pkce): added pkce support

### LINKED ISSUE #76: Implement PKCE
See https://snyk.io/vuln/npm:oauth2-server
https://github.com/oauthjs/node-oauth2-server/pull/452

### LINKED ISSUE #89: [meta] list of original project pr
Hi,

this is a list of original project pr still open, to be analyzed and possibly integrated in this code base.

* [x] 1 [2.x support: oauth.token promise, returns token object instead of sending response - 424](https://github.com/oauthjs/node-oauth2-server/pull/424)
    * This is referring to a very old version, no more relevant
* [x] 2 [PKCE - 452](https://github.com/oauthjs/node-oauth2-server/pull/452)
    * Addressed in issue #76, so we do not need to track it here 
* [x] 3 [Don't fail on vschar check on state if it is empty and allowEmptyState=1 - 473 ](https://github.com/oauthjs/node-oauth2-server/pull/473)
    * see https://github.com/node-oauth/node-oauth2-server/pull/93
* [x] 4 [Implement validateRedirectUri model function - 482](https://github.com/oauthjs/node-oauth2-server/pull/482)
    * This is a feature-adding pr. Since it seems fine, I just added it in pr #97. The code is slightly changed w.r.t. original pr (just a little cleaner imho).
* [x] 5 [ Update extension-grants.rst with example - 530](https://github.com/oauthjs/node-oauth2-server/pull/530)
    * Fixed by pr #92
* [x] 6 [Allow "allowed" parameter to be fetched from body - 532](https://github.com/oauthjs/node-oauth2-server/pull/532)
    * see https://github.com/node-oauth/node-oauth2-server/pull/94
* [x] 7 [Pass scope to model on refresh_token request - 540](https://github.com/oauthjs/node-oauth2-server/pull/540)
    * The original pr is fine w.r.t. the RFC6749 Section 6. I believe the missing part is to ensure the passed `scope` does not contains non-validated scopes, and this missing check could lead to potential security risks. So I suggest to implement a check to ensure the passed request scope object is a "subset" of validated scopes (or, as alternative, a method which removes invalid scopes).
* [x] 8 [Update README.md - 556](https://github.com/oauthjs/node-oauth2-server/pull/556)
    * Fixed by pr #91
* [x] 9 [rokenToken is not required if alwaysIssueNewRefreshToken = false - 558](https://github.com/oauthjs/node-oauth2-server/pull/558)
    * The pr is right, w.r.t. the rest of the code base. So it is is just a little improvement, to make the `model.revokeToken()` method implementation optional. So the pr could be accepted, even if it is not strictly useful (it is possible to implement the method with an empty body, or maybe with a default implementation inside a "model" base class). Therefore I am not proposing a pr for this.
* [x] 10 [Add return statements for more clarity - 561](https://github.com/oauthjs/node-oauth2-server/pull/561)
    * Pr proposed: #98. This is a simple "doc" improvement. The original pr states that is for clarity, but it is not: the doc examples are actually bugged, so this pr is required.
* [x] 11 [RFC 6749#4.1.2.1 - fix order of error handling - 565](https://github.com/oauthjs/node-oauth2-server/pull/565)
    * Pr proposed: #112. The original pr fixes the order of some checks, according with the standard, which seems fine to me.
* [x] 12 [The object returned by the response.getHeaders() method does not prot… - 567](https://github.com/oauthjs/node-oauth2-server/pull/567)
    * This pr proposes a fix already present in the code. So it is useless. 
* [x] 13 [Update authorize-handler.js - 576](https://github.com/oauthjs/node-oauth2-server/pull/576)
    * This pr is an alternative implementation of pr https://github.com/oauthjs/node-oauth2-server/pull/649. See comments below (point number 20). 
* [x] 14 [v5: add codecoverage and upgrade packages - 610](https://github.com/oauthjs/node-oauth2-server/pull/610)
    * This targets branch `v5-dev` (typescript). Probably no more useful.
* [x] 15 [fix: Bearer regular expression matching in authenticate handler - 614](https://github.com/oauthjs/node-oauth2-server/pull/614)
    * The original pr is fine. Proposed the pr #105 
* [x] 16 [Update refresh-token-grant-type.js - 615](https://github.com/oauthjs/node-oauth2-server/pull/615)
    * This pr is wrong. The change introduce a bug. 
* [x] 17 [Passing validated scope into generateAccessToken. - 620](https://github.com/oauthjs/node-oauth2-server/pull/620)
    * The pr targets branch `v5-dev`, and claims it is already fixed in other branches. So it is useless.
* [x] 18 [Set WWW-Authenticate header for invalid requests - 646](https://github.com/oauthjs/node-oauth2-server/pull/646)
    * The original pr is right, so I just re-proposed it here: #96 
* [ ] 19 [Remove scope validation - 647](https://github.com/oauthjs/node-oauth2-server/pull/647)
    * Proposed pr #103 
    * The implementation is slightly different from the original pr. Details given in the pr description 
* [x] 20 [Fix user denial and missing state parameter on error redirect - 649](https://github.com/oauthjs/node-oauth2-server/pull/649)
    * This is a different implementation of point 13 
    * Propose pr #99
* [x] 21 [Update refresh-token-grant-type.js - 657](https://github.com/oauthjs/node-oauth2-server/pull/657)
    * The pr does not looks correct. Checking also the `v5-dev` branch, the client object seems to have a `id` member, and not a `clientId` member. Written into the original pr to have feedback.
* [x] 22 [Add PKCE support - 658](https://github.com/oauthjs/node-oauth2-server/pull/658)
    * Addressed in issue #76, so we do not need to track it here
* [x] 23 [Add PKCE support (for v5-dev Typescript) - 659](https://github.com/oauthjs/node-oauth2-server/pull/659)
    * Addressed in issue #76, so we do not need to track it here
* [ ] 24 [Fix model spec to use code.authorizationCode instead of code.code - 660](https://github.com/oauthjs/node-oauth2-server/pull/660)
    * The original pr just upgrades the doc, according with the original `dev` branch. Even if the pr seems not strictly useful right now, it shows that the original `dev` branch could contain some interesting improvements, worth to merge in this code base. To be further analyzed.
* [x] 25 [dependencies updated #662](https://github.com/oauthjs/node-oauth2-server/pull/662)
    * This targets branch `v5-dev` (typescript). Probably no more useful.
* [x] 26 [Bump statuses from 1.5.0 to 2.0.1 - 670](https://github.com/oauthjs/node-oauth2-server/pull/670)
    * No more relevant: package has been removed
* [x] 27 [Update token-util.js - 678](https://github.com/oauthjs/node-oauth2-server/pull/678)
    * Already fixed
* [x] 28 [Bump mocha from 5.2.0 to 8.4.0 - 689](https://github.com/oauthjs/node-oauth2-server/pull/689)
    * Fixed by the following n.32
* [x] 29 [Bump sinon from 7.5.0 to 11.1.1 - 694](https://github.com/oauthjs/node-oauth2-server/pull/694)
    * Fixed by the following n.33
* [x] 30 [Bump jshint from 2.13.0 to 2.13.1 - 708](https://github.com/oauthjs/node-oauth2-server/pull/708)
    * No more relevant: jshint has been replaced with eslint
* [ ] 31 [Only require 'redirect_uri' in token/authorization_code when one was provided in authorize/code - 709](https://github.com/oauthjs/node-oauth2-server/pull/709)
    * The original pr question is correct. But I am wondering whether the proposed code is fine. In fact I am not 100% sure about the `getRedirectUri()` method implementation. Probably, just removing `|| client.redirectUris[0]` will fix all the issues. But I am not completely sure about this: I do not know the standard and the code so well...
* [x] 32 [Bump mocha from 5.2.0 to 9.1.3 - 716](https://github.com/oauthjs/node-oauth2-server/pull/716)
    * Fixed by pr #90
* [x] 33 [Bump sinon from 7.5.0 to 12.0.1 - 718](https://github.com/oauthjs/node-oauth2-server/pull/718)
    * Already fixed

The list will be upgraded to track the ongoing process of integrating the pr's.

Regards.

### PR DESCRIPTION
**Pull request by jankapunkt**
## Summary

This implements PKCE support (related: #76) similar to https://github.com/oauthjs/node-oauth2-server/pull/658

However, this did not 1:1 copy the implementation and rather modified it.

## Linked issue(s)

#76

## Involved parts of the project

- `AuthorizationCodeGrantType`
- `TokenHandler`

In the original PR there was also `CodeResponseType` covered. However, they targetetd `oauthjs:dev` and the content of this file differs from ours (and from `3.1.1`: )

- CodeResponseType in `oauthjs:3.1.1`: https://github.com/oauthjs/node-oauth2-server/blob/v3.1.1/lib/response-types/code-response-type.js (this is the same as ours on `development`, right?)
- CodeResponseType in `oauthjs:dev`: https://github.com/oauthjs/node-oauth2-server/blob/dev/lib/response-types/code-response-type.js
- CodeResponseType in their PR: https://github.com/oauthjs/node-oauth2-server/blob/1dd54a4cab61c9771bdeeba1c74afc5a4e5c2b65/lib/response-types/code-response-type.js

This is open to debate as I am not 100% sure if the implementation for `CodeResponseType` is even required anymore.

## Added tests?

Yes

## OAuth2 standard

[RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636)

## Reproduction

Checkout this branch and run the tests.

## Important

This needs to be tested very well, including integration with our own implementations!

### CODE REVIEW

**Comment by jankapunkt** [comment]
@Uzlopak do you mind checking this out?

**Comment by jankapunkt** [comment]
Someone found the time to test this?

**Comment by jorenvandeweyer** [comment]
I'll test and review this in the coming days.

**Review by jorenvandeweyer — COMMENTED** [review]
Everything looks fine, but maybe we should add some documentation?

**Comment by jankapunkt** [comment]
Definitely, otherwise it will be really hard to set it up

**Comment by jorenvandeweyer** [comment]
We can already merge it if you want since we are not hosting our documentation at this moment.

(We also need to update the typescript definitions.)

**Comment by jankapunkt** [comment]
> We can already merge it if you want since we are not hosting our documentation at this moment.

We don't? Well, I will open an issue to use GitHub pages for that, I can also provide the PR for it.

> (We also need to update the typescript definitions.)

For this PKCE? Can you add this to the PR, please? I have no TS experience.

**Comment by jirehnimes** [comment]
Any updates about this?

**Comment by jankapunkt** [comment]
@jirehnimes it's basically implemented but needs further testing by people who use it in their workflow and a proper review. If you'd like to test it please let me know so I can advice you, if required.

**Comment by jirehnimes** [comment]
> @jirehnimes it's basically implemented but needs further testing by people who use it in their workflow and a proper review. If you'd like to test it please let me know so I can advice you, if required.

My bad. Thanks for the fast response!

**Comment by jankapunkt** [comment]
@martinssonj would you mind adding a final review to this one? I will follow this week and then we can merge this and release 4.3.0

File: test/integration/handlers/authorize-handler_test.js [hunk 1]
```diff
@@ -1301,4 +1301,44 @@ describe('AuthorizeHandler integration', function() {
...
+      codeChallengeMethod.should.equal('S256');
+    });
+
+    it('should get default code challenge method plain if missing', function() {
+      const model = {
+        getAccessToken: function() {},
+        getClient: function() {},
+        saveAuthorizationCode: function() {}
+      };
+      const handler = new AuthorizeHandler({ authorizationCodeLifetime: 120, model: model });      const request = new Request({ body: {}, headers: {}, method: {}, query: {} });
```
Comment (martinssonj) [inline test/integration/handlers/authorize-handler_test.js:1324]:
```suggestion
      const handler = new AuthorizeHandler({ authorizationCodeLifetime: 120, model: model });      
      const request = new Request({ body: {}, headers: {}, method: {}, query: {} });
```

File: test/integration/handlers/authorize-handler_test.js [hunk 2]
```diff
@@ -1301,4 +1301,44 @@ describe('AuthorizeHandler integration', function() {
...
+  });
+
+  describe('getCodeChallenge()', function() {
+    it('should get code challenge', function() {
+      const model = {
+        getAccessToken: function() {},
+        getClient: function() {},
+        saveAuthorizationCode: function() {}
+      };
+      const handler = new AuthorizeHandler({ authorizationCodeLifetime: 120, model: model });      const request = new Request({ body: {code_challenge: 'challenge'}, headers: {}, method: {}, query: {} });
```
Comment (martinssonj) [inline test/integration/handlers/authorize-handler_test.js:1338]:
```suggestion
      const handler = new AuthorizeHandler({ authorizationCodeLifetime: 120, model: model });      
      const request = new Request({ body: {code_challenge: 'challenge'}, headers: {}, method: {}, query: {} });
```

**Comment by martinssonj** [comment]
> @martinssonj would you mind adding a final review to this one? I will follow this week and then we can merge this and release 4.3.0

I will test it out in my project a final time tomorrow. Otherwise, it looks good to me.

**Review by martinssonj — APPROVED** [review]

**Comment by martinssonj** [comment]
Hi @jankapunkt. What do you think about merging this one? Let me know if there is anything I can assist with for this PR.

**Comment by jankapunkt** [comment]
Hey @martinssonj thanks for the reminder. I will add a review this weekend, promise (.all) 😅

**Comment by jankapunkt** [comment]
@martinssonj while reading again through RFC7636 I found that 4.4.1 missed to throw an error when an unknown (but not undefined) code challenge method, such as `'foo'` is in the request. I added this in the latest commit.

I am reading through a few edge cases and see if we have them covered, too.

**Review by martinssonj — APPROVED** [review]
Nice that you found this.

**Review by jankapunkt — COMMENTED** [review]
@martinssonj from my end it's all good now, would you mind adding a final review on this (I added ~~two~~ one more test) as I can't review my own pull request by our ruleset.

After that I will merge and release 4.3.0

@jorenvandeweyer @Uzlopak  @HappyZombies feel free to comment / review, too if you like

**Review by martinssonj — APPROVED** [review]
Looks good.

**Comment by jankapunkt** [comment]
@martinssonj @jirehnimes the release is already available https://www.npmjs.com/package/@node-oauth/oauth2-server/v/4.3.0 

The master branch is not updated yet as I have to fiddle with some legacy stuff in our CI

**Comment by kehers** [comment]
The current implementation expects `code_challenge` and `code_challenge_method` as POST parameters and not GET ([authorize-handler:379](https://github.com/node-oauth/node-oauth2-server/blob/c993eb5a700f81fe204283b3428e0742015f9b8d/lib/handlers/authorize-handler.js#L379) and [authorize-handler:390](https://github.com/node-oauth/node-oauth2-server/blob/c993eb5a700f81fe204283b3428e0742015f9b8d/lib/handlers/authorize-handler.js#L390)). Curious if this should be so.

**Comment by jankapunkt** [comment]
> The current implementation expects `code_challenge` and `code_challenge_method` as POST parameters and not GET ([authorize-handler:379](https://github.com/node-oauth/node-oauth2-server/blob/c993eb5a700f81fe204283b3428e0742015f9b8d/lib/handlers/authorize-handler.js#L379) and [authorize-handler:390](https://github.com/node-oauth/node-oauth2-server/blob/c993eb5a700f81fe204283b3428e0742015f9b8d/lib/handlers/authorize-handler.js#L390)). Curious if this should be so.

Hi @kehers, the [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636) does not explicitly state which of both methods to use. Is there an explicit issue in your setup by using POST?

// cc @martinssonj

**Comment by martinssonj** [comment]
It looks like GET should be supported by `authorize` from looking at this: https://datatracker.ietf.org/doc/html/rfc6749#section-3-1 and getResponseType is supporting query parameters so I think code challenge methods should as well.

**Pull request merged**